python/pyiceberg/table/__init__.py

```diff
@@ -145,12 +152,22 @@ def current_snapshot(self) -> Optional[Snapshot]:
             return self.snapshot_by_id(snapshot_id)
         return None
 
-    def snapshot_by_id(self, snapshot_id: int) -> Optional[Snapshot]:
-        """Get the snapshot of this table with the given id, or None if there is no matching snapshot."""
+    def snapshot_by_id(self, snapshot_id: int) -> Snapshot:
+        """Get the snapshot of this table with the given id.
+
+        Args:
+            snapshot_id: The id of the snapshot to lookup in the table
+
+        Returns:
+            The snapshot that corresponds to snapshot_id
+
+        Raises:
+            ValueError: If the snapshot cannot be found
+        """
         try:
             return next(snapshot for snapshot in self.metadata.snapshots if snapshot.snapshot_id == snapshot_id)
-        except StopIteration:
-            return None
+        except StopIteration as e:
+            raise ValueError(f"Snapshot id not found in table: {snapshot_id}") from e
```

Comment:
Should we just index these in a dict and let it raise KeyError?

Reply:
The `next(...)` pattern is more efficient as it will stop the iteration once the snapshot has been found. I'm also happy to materialize it as a `dict` that we cache.

Reply:
Yeah, I was thinking a dict that we cache. Not a big deal, but we'll probably want to do that eventually.

Reply:
Yes, I've changed it into exactly that: https://github.com/apache/iceberg/pull/6933/files#diff-893817e2d326eb9fad68cc96f7742b3ca654a2e05355f9d80ed59886b36ac121R170-R177